Arm/State.lean

```diff
@@ -662,3 +662,296 @@ theorem write_mem_irrelevant :
+and `write_mem`'. These operate directly on the memory, rather than the `ArmState`.
+We prove their equivalence to the existing definitions (`read_mem_eq_read_mem'`, `write_mem_eq_write_mem'`).
+
+Furthermore, we define equivalences to `BitVec.extractLsByte` to ease reasoning about byte-level manipulation.
+As an easy corollary, we get `getLsb` theorems on these in order to allow `omega` based reasoning about bit-level values of memory.
+
+We then build lemmas that allow simplification of the proof state given the new `mem_subset'` and `mem_separate'` assumptions.
+In total, this gives us automation to simplify theorems about memory (non)-interference.
+-/
+
+abbrev Memory := Store (BitVec 64) (BitVec 8)
+
+/--
+A variant of `read_mem` that directly talks about writes to memory, instead of over the entire `ArmState`,
+-/
+def read_mem' (addr : BitVec 64) (m : Memory) : BitVec 8 :=
+  read_store addr m
+
+theorem read_mem_eq_read_mem' : read_mem addr s = read_mem' addr s.mem := rfl
+
+@[simp]
+theorem getLsb_read_mem' : (read_mem' addr s).getLsb i = (s addr).getLsb i := rfl
+
+def read_mem_bytes' (n : Nat) (addr : BitVec 64) (s : Memory) : BitVec (n * 8) :=
+  match n with
+  | 0 => 0#0
+  | n' + 1 =>
+    let byte := read_mem' addr s
+    let rest := read_mem_bytes' n' (addr + 1#64) s
+    have h: n' * 8 + 8 = (n' + 1) * 8 := by simp_arith
+    BitVec.cast h (rest ++ byte)
+
+theorem read_mem_bytes_eq_read_mem_bytes' (s : ArmState) :
+    read_mem_bytes n addr s = read_mem_bytes' n addr s.mem := by
+  induction n generalizing addr s
+  case zero => simp [read_mem_bytes, read_mem_bytes']
+  case succ n' ih =>
+    simp [read_mem_bytes, read_mem_bytes', read_mem_eq_read_mem', ih]
+
+@[simp]
+theorem read_mem_bytes'_zero_eq : read_mem_bytes' 0 addr s = 0#0 := rfl
+
+theorem read_mem_bytes'_succ_eq :
+  read_mem_bytes' (n' + 1) addr s = ((read_mem_bytes' n' (addr + 1) s) ++ read_mem' addr s).cast (by omega) := rfl
+
+theorem getLsb_read_mem_bytes' {n i : Nat} {addr : BitVec 64} {s : Memory} (hn : n ≤ 2^64) :
+    (read_mem_bytes' n addr s).getLsb i =
+    (decide (i < n * 8) && (s (addr + BitVec.ofNat 64 (i / 8))).getLsb (i % 8)) := by
+  induction n generalizing i addr s
+  case zero =>
+    simp
+  case succ n' ih =>
+    simp only [read_mem_bytes'_succ_eq, BitVec.ofNat_eq_ofNat, BitVec.getLsb_cast,
+      BitVec.getLsb_append, getLsb_read_mem']
+    rw [Nat.succ_mul]
+    by_cases h₁ : (i < 8)
+    · simp only [h₁, decide_True, cond_true, show i < n' * 8 + 8 by omega, Bool.true_and]
+      have hdiv : i / 8 = 0 :=  Nat.div_eq_of_lt h₁
+      rw [hdiv]
+      simp only [BitVec.add_zero]
+      have hmod : i % 8 = i := Nat.mod_eq_of_lt h₁
+      simp [hmod]
+    · simp only [h₁, decide_False, cond_false]
+      rw [ih]
+      by_cases h₂ : i - 8 < n' * 8
+      · simp only [h₂, decide_True, Bool.true_and, show (i < n' * 8 + 8) by omega]
+        have hi' : ∃ i', i = i' + 8 := by
+          apply Classical.byContradiction
+          intros h
```

Comment:
```suggestion
          by_contra h
```

`by_contra` is not a mathlib-ism, is it?

Reply:
Sorry Alex, it's in batteries :(